spec.html

```diff
@@ -32093,6 +32093,12 @@ <h1>Array.prototype.flat ( [ _depth_ ] )</h1>
         <emu-clause id="sec-flattenintoarray" aoid="FlattenIntoArray">
           <h1>FlattenIntoArray ( _target_, _source_, _sourceLen_, _start_, _depth_ [ , _mapperFunction_, _thisArg_ ] )</h1>
           <emu-alg>
+            1. Assert: Type(_target_) is Object.
+            1. Assert: Type(_source_) is Object.
+            1. Assert: _sourceLen_ is an integer Number &ge; 0.
+            1. Assert: _start_ is an integer Number &ge; 0.
+            1. Assert: _depth_ is an integer Number, *+&infin;*, or *-&infin;*.
+            1. Assert: if _mapperFunction_ is present, ! IsCallable(_mapperFunction_) is *true*, _thisArg_ is present, and _depth_ is *1*.
```

Comment:
```suggestion
            1. Assert: If _mapperFunction_ is present, then ! IsCallable(_mapperFunction_) is *true* and _thisArg_ is present.
```
- Uppercase "I" in "If" because the spec has a strong preference for a capital after "Assert:".
- Insert "then" because otherwise there's a potential garden-path.
(E.g., `If <A>, <B>, ...` could be the start of
`If <A>, <B>, <C>, and <D>.` (meaning `If <A> then <B> and <C> and <D>.`) 
or
`If <A>, <B>, and <C>, <D>.` (meaning `If <A> and <B> and <C> then <D>.`)
It isn't an ambiguity per se, because the position of the "and" tells you where a compound condition ends, but the reader shouldn't have to look ahead and figure it out.)

- Drop the `_depth_ is *1*` condition because, while it happens to be true in the current spec, it's not needed for the correct functioning of this operation. (Consider a variant of `flatMap` that's equivalent to a `map` followed by a `flat` of specified (or infinite) depth: its semantics should be able to invoke `FlattenIntoArray`, and that would work fine, *except* that this `_depth_ is *1*` assertion would fail.)

Reply:
I added the depth assertion because the recursive call doesn’t preserve the mapperFunction or thisArg, so without modification it couldn’t work with that variant anyways.

If that modification is made separately, the depth 1 part of the assertion would thus be removed.

Reply:
No, I'm saying just a top-level map, followed by X levels of flattening, so the recursive call is just right for that, no modification needed.

Reply:
Maybe we’re talking past each other - if flatMao took a depth (or flat took a mapper), and passed it into the current FlattenIntioArray, it would not work properly, because the mapper wouldn’t be invoked for nested items.

Reply:
Whether it "works properly" depends on what you think it should do.
- If you're coming from lodash, you would say that it works properly.
- Or consider the quick description of `flatMap`: that it's like a call to `map` followed by a call to `flat`. So if `flatMap` took a depth, then a reasonable expectation would be that it's like a call to `map` followed by a call to `flat(depth)`. So from that point of view, it would also work properly.

I agree that there might be situations in which you'd want the mapper to be applied recursively (whenever it generates an array), and I agree that `FlattenIntoArray` doesn't do that. But the fact that it doesn't support *that* functionality isn't a reason to say that it can't handle a *different* (reasonable, useful) functionality.

Reply:
Fair point. However, the spec currently does not call this operation with both a mapper function and a depth that isn’t 1. So I think the assertion is still appropriate at the moment.

Reply:
While it's true that the spec only calls `FlattenIntoArray` in ways that satisfy the assertion, those aren't necessarily the only calls -- it's an open spec, and downstream specs can apparently re-use whatever they find useful. An implementation/host might provide (say) `flatMapDepth` and specify it in terms of ECMA-262 operations. If they respect the assertion in question, they'd have to construct their own version of `FlattenIntoArray` that's identical except for omitting the `_depth_ is *1*` condition. Or (more likely?), they'd just use `FlattenIntoArray`, acknowledge the assertion failure, and tell the reader to pretend it doesn't happen. Neither solution is great. So why create unnecessary problems for downstream?

(There's also the case that a future edition of ECMA-262 might wish to add such a function. That's less compelling, because whoever adds the function can remove the `_depth_ is *1*` condition at the same time. But it does still assume that people recognize that the assertion is unnecessary.)

Reply:
I think it's the appropriate thing to do to restrict the possible uses of this abstract operation to what 262 does (modulo known external usages of it) - abstract operations are not a utility library.

Reply:
> I think it's the appropriate thing to do to restrict the possible uses of this abstract operation to what 262 does (modulo known external usages of it) 

Okay. I disagree, but I'm not going to argue the point further.

> abstract operations are not a utility library.

That's... interesting. I think that deserves a separate issue.